sycl/source/detail/graph_impl.cpp

```diff
@@ -1027,141 +1042,297 @@ exec_graph_impl::~exec_graph_impl() {
-      CGData.MEvents = BackupCGDataMEvents;
-    }
-
-    for (auto const &DepPartition : CurrentPartition->MPredecessors) {
-      CGData.MEvents.push_back(PartitionsExecutionEvents[DepPartition]);
-    }
-
-    auto CommandBuffer = CurrentPartition->MCommandBuffers[Queue.get_device()];
-
-    if (CommandBuffer) {
-      for (std::vector<sycl::detail::EventImplPtr>::iterator It =
-               MExecutionEvents.begin();
-           It != MExecutionEvents.end();) {
-        EventImplPtr &Event = *It;
-        if (!Event->isCompleted()) {
-          const std::vector<EventImplPtr> &AttachedEventsList =
-              Event->getPostCompleteEvents();
-          CGData.MEvents.reserve(CGData.MEvents.size() +
-                                 AttachedEventsList.size() + 1);
-          CGData.MEvents.push_back(Event);
-          // Add events of the previous execution of all graph partitions.
-          CGData.MEvents.insert(CGData.MEvents.end(),
-                                AttachedEventsList.begin(),
-                                AttachedEventsList.end());
-          ++It;
-        } else {
-          // Remove completed events
-          It = MExecutionEvents.erase(It);
-        }
+  }
+}
+
+EventImplPtr
+exec_graph_impl::enqueuePartitions(sycl::detail::queue_impl &Queue,
+                                   sycl::detail::CG::StorageInitHelper &CGData,
+                                   bool IsCGDataSafeForSchedulerBypass,
+                                   bool EventNeeded) {
+
+  // If EventNeeded is true, this vector is used to keep track of dependencies
+  // for the returned event. This is used when the graph has multiple end nodes
+  // which cannot be tracked with a single scheduler event.
+  std::vector<EventImplPtr> PostCompleteDependencies;
+
+  // This variable represents the returned event. It will always be nullptr if
+  // EventNeeded is false.
+  EventImplPtr SignalEvent;
+
+  // CGData.MEvents gets cleared after every partition enqueue. If we need the
+  // original events, a backup needs to be created now. This is only needed when
+  // the graph contains more than one root partition.
+  std::vector<detail::EventImplPtr> BackupCGDataEvents;
+  if (MRootPartitions.size() > 1) {
+    BackupCGDataEvents = CGData.MEvents;
+  }
+
+  for (auto &Partition : MPartitions) {
+
+    if (Partition->MPredecessors.empty() && CGData.MEvents.empty()) {
+      // If this is a root partition and CGData has been cleared already, we
+      // need to restore it so that the partition execution waits for the
+      // dependencies of this graph execution.
+      CGData.MEvents = BackupCGDataEvents;
+    } else {
+      // Partitions can have multiple dependencies from previously executed
+      // partitions. To enforce this ordering, we need to add these dependencies
+      // to CGData.
+      for (auto &Predecessor : Partition->MPredecessors) {
+        CGData.MEvents.push_back(Predecessor.lock()->MEvent);
```

Comment:
If you're only using `std::weak_ptr` to avoid memory leaks, but not for the actual memory management (i.e., entries are never `expired` and `lock` never fails), then why is that a `weak_ptr` and not a raw pointer?

Reply:
I was mostly following what seemed to be the standard in the current implementation. There is other examples where we are using weak_ptr for similar situations.

But that's fair enough. There is no reason to use `weak_ptr` in this case and I have updated the implementation to use raw ptr.

Additionally, as a follow up to the refactoring you are doing in the sycl runtime, I also created an internal task for our team to go through the `graph` codebase and update it to use raw pointers where lifetime tracking is not needed.